grpc-context-utils/src/main/java/org/hypertrace/core/grpcutils/context/DefaultContextualKey.java

```diff
@@ -0,0 +1,73 @@
+package org.hypertrace.core.grpcutils.context;
+
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+class DefaultContextualKey<T> implements ContextualKey<T> {
+  private final RequestContext context;
+  private final T data;
+  private final Map<String, String> meaningfulContextHeaders;
+
+  DefaultContextualKey(RequestContext context, T data) {
```

Comment:
Do we always have to pass in some data to make the request context as a cache key? Can we also have a constructor, which just takes in the `RequestContext` as an argument?

Reply:
I am a bit confused why do we need `data` (_which I believe is an input to the GRPC function calls_) as a parameter for the cache key

Basically, why should the contextual key change on a different `data`?

Reply:
>Do we always have to pass in some data to make the request context as a cache key? Can we also have a constructor, which just takes in the RequestContext as an argument?

We don't always need to use data, there's a no arg version (this is always built off the context) https://github.com/hypertrace/java-grpc-utils/pull/16/files#diff-ef190554e9c2549e8421789de3b00e6c5fae5294934d883251699dff75676817R87 )

I'll create flavors of the function execution methods that don't provide data, too.


> I am a bit confused why do we need data (which I believe is an input to the GRPC function calls) as a parameter for the cache key
>
> Basically, why should the contextual key change on a different data?

So I attempt to convey the distinction by calling a Contextual key rather than a Context key - that is, it's a key that accounts for the context rather than a key for the context. Since the former is a superset of the latter, it felt more flexible to me as it can both handle cases where we solely want a per-context key, but also cases where we, for example, want to cache something like an entity by id.

Reply:
Updated the first part.